File: tests/api/java/src/main/java/opencall/Router.java
package opencall;

import java.time.Instant;
import java.time.LocalDate;
import java.util.*;

/**
 * Envelope dispatcher for the OpenCALL Todo API.
 */
public final class Router {

    @SuppressWarnings("unchecked")
    public static Map<String, Object> handleCall(
        Map<String, Object> envelope,
        String authHeader,
        Operations.MediaFile mediaFile
    ) {
        Map<String, Object> ctx = envelope.get("ctx") instanceof Map
            ? (Map<String, Object>) envelope.get("ctx")
            : new LinkedHashMap<>();

        String requestId = ctx.get("requestId") instanceof String
            ? (String) ctx.get("requestId")
            : UUID.randomUUID().toString();

        String sessionId = ctx.get("sessionId") instanceof String
            ? (String) ctx.get("sessionId")
            : null;

        LinkedHashMap<String, Object> base = new LinkedHashMap<>();
        base.put("requestId", requestId);
        if (sessionId != null) {
            base.put("sessionId", sessionId);
        }

        // Validate op is present and a string
        Object opObj = envelope.get("op");
        if (opObj == null || !(opObj instanceof String) || ((String) opObj).isEmpty()) {
            LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
            body.put("state", "error");
            LinkedHashMap<String, Object> error = new LinkedHashMap<>();
            error.put("code", "INVALID_REQUEST");
            error.put("message", "Missing or invalid 'op' field");
            body.put("error", error);

            LinkedHashMap<String, Object> result = new LinkedHashMap<>();
            result.put("status", 400);
            result.put("body", body);
            return result;
        }

        String op = (String) opObj;

        // Look up operation
        Operations.OperationEntry operation = Operations.OPERATIONS.get(op);
        if (operation == null) {
            LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
            body.put("state", "error");
            LinkedHashMap<String, Object> error = new LinkedHashMap<>();
            error.put("code", "UNKNOWN_OP");
            error.put("message", "Unknown operation: " + op);
            body.put("error", error);

            LinkedHashMap<String, Object> result = new LinkedHashMap<>();
            result.put("status", 400);
            result.put("body", body);
            return result;
        }

        // Deprecated check -- past sunset date means 410
        if (operation.deprecated && operation.sunset != null) {
            LocalDate sunsetDate = LocalDate.parse(operation.sunset);
            if (LocalDate.now().isAfter(sunsetDate)) {
                LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                body.put("state", "error");
                LinkedHashMap<String, Object> error = new LinkedHashMap<>();
                error.put("code", "OP_REMOVED");
                error.put("message", "Operation " + op + " has been removed");
                LinkedHashMap<String, Object> cause = new LinkedHashMap<>();
                cause.put("removedOp", op);
                cause.put("replacement", operation.replacement);
                error.put("cause", cause);
                body.put("error", error);

                LinkedHashMap<String, Object> result = new LinkedHashMap<>();
                result.put("status", 410);
                result.put("body", body);
                return result;
            }
        }

        // Auth check
        if (operation.authScopes != null && !operation.authScopes.isEmpty()) {
            Auth.AuthResult authResult = Auth.validateAuth(authHeader, operation.authScopes);
            if (!authResult.valid()) {
                LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                body.put("state", "error");
                LinkedHashMap<String, Object> error = new LinkedHashMap<>();
                error.put("code", authResult.code());
                error.put("message", authResult.message());
                body.put("error", error);

                LinkedHashMap<String, Object> result = new LinkedHashMap<>();
                result.put("status", authResult.status());
                result.put("body", body);
                return result;
            }
        }

        // Check idempotency store for side-effecting ops
        String idempotencyKey = ctx.get("idempotencyKey") instanceof String
            ? (String) ctx.get("idempotencyKey")
            : null;

        if (operation.sideEffecting && idempotencyKey != null) {
            Object cached = Operations.getIdempotencyStore().get(idempotencyKey);
            if (cached != null) {
                return (Map<String, Object>) cached;
            }
        }

        // Execute handler
        Object rawArgs = envelope.get("args");
        if (rawArgs == null) rawArgs = new LinkedHashMap<>();

        try {
            // Stream operations
            if ("stream".equals(operation.executionModel) && operation.streamHandler != null) {
                Map<String, Object> streamResult = operation.streamHandler.apply(rawArgs);
                if (!Boolean.TRUE.equals(streamResult.get("ok"))) {
                    LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                    body.put("state", "error");
                    body.put("error", streamResult.get("error"));

                    LinkedHashMap<String, Object> result = new LinkedHashMap<>();
                    result.put("status", 200);
                    result.put("body", body);
                    return result;
                }

                String streamSessionId = (String) streamResult.get("sessionId");
                LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                body.put("state", "streaming");
                LinkedHashMap<String, Object> stream = new LinkedHashMap<>();
                stream.put("transport", "wss");
                stream.put("location", "/streams/" + streamSessionId);
                stream.put("sessionId", streamSessionId);
                stream.put("encoding", "json");
                stream.put("expiresAt", Instant.now().plusSeconds(3600).getEpochSecond());
                body.put("stream", stream);

                LinkedHashMap<String, Object> result = new LinkedHashMap<>();
                result.put("status", 202);
                result.put("body", body);
                return result;
            }

            // Async operations
            if ("async".equals(operation.executionModel) && operation.asyncHandler != null) {
                Map<String, Object> asyncResult = operation.asyncHandler.apply(rawArgs, requestId);
                if (!Boolean.TRUE.equals(asyncResult.get("ok"))) {
                    LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                    body.put("state", "error");
                    body.put("error", asyncResult.get("error"));

                    LinkedHashMap<String, Object> result = new LinkedHashMap<>();
                    result.put("status", 200);
                    result.put("body", body);
                    return result;
                }

                LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                body.put("state", "accepted");
                body.put("retryAfterMs", 100);
                body.put("expiresAt", Instant.now().plusSeconds(3600).getEpochSecond());

                LinkedHashMap<String, Object> result = new LinkedHashMap<>();
                result.put("status", 202);
                result.put("body", body);
                return result;
            }

            // Sync operations (including media handler)
            Map<String, Object> handlerResult;
            if (operation.acceptsMedia) {
                handlerResult = Operations.callAttachHandler(rawArgs, mediaFile);
            } else {
                handlerResult = operation.handler.apply(rawArgs);
            }

            LinkedHashMap<String, Object> response;
            if (Boolean.TRUE.equals(handlerResult.get("ok"))) {
                LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                body.put("state", "complete");
                body.put("result", handlerResult.get("result"));

                response = new LinkedHashMap<>();
                response.put("status", 200);
                response.put("body", body);
            } else {
                // Domain error -- HTTP 200
                LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
                body.put("state", "error");
                body.put("error", handlerResult.get("error"));

                response = new LinkedHashMap<>();
                response.put("status", 200);
                response.put("body", body);
            }

            // Store for idempotency
            if (operation.sideEffecting && idempotencyKey != null) {
                Operations.getIdempotencyStore().put(idempotencyKey, response);
            }

            return response;

        } catch (Operations.ValidationError err) {
            LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
            body.put("state", "error");
            LinkedHashMap<String, Object> error = new LinkedHashMap<>();
            error.put("code", "VALIDATION_ERROR");
            error.put("message", err.getMessage());
            body.put("error", error);

            LinkedHashMap<String, Object> result = new LinkedHashMap<>();
            result.put("status", 400);
            result.put("body", body);
            return result;

        } catch (Operations.ServerError err) {
            LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
            body.put("state", "error");
            LinkedHashMap<String, Object> error = new LinkedHashMap<>();
            error.put("code", err.code);
            error.put("message", err.getMessage());
            body.put("error", error);

            LinkedHashMap<String, Object> result = new LinkedHashMap<>();
            result.put("status", err.statusCode);
            result.put("body", body);
            return result;

        } catch (Exception err) {
            LinkedHashMap<String, Object> body = new LinkedHashMap<>(base);
            body.put("state", "error");
            LinkedHashMap<String, Object> error = new LinkedHashMap<>();
            error.put("code", "INTERNAL_ERROR");
            error.put("message", err.getMessage() != null ? err.getMessage() : "Unknown error");
            body.put("error", error);

            LinkedHashMap<String, Object> result = new LinkedHashMap<>();
            result.put("status", 500);
            result.put("body", body);
            return result;
        }
    }
}

File: tests/api/go/router.go
package main

import (
	"fmt"
	"time"
)

func handleCall(envelope map[string]interface{}, authHeader string, mf *mediaFile) (result map[string]interface{}) {
	// Extract requestId and sessionId from ctx
	requestID := newUUID()
	var sessionID string
	hasSessionID := false

	if ctx, ok := envelope["ctx"].(map[string]interface{}); ok {
		if rid, ok := ctx["requestId"].(string); ok && rid != "" {
			requestID = rid
		}
		if sid, ok := ctx["sessionId"].(string); ok && sid != "" {
			sessionID = sid
			hasSessionID = true
		}
	}

	base := map[string]interface{}{
		"requestId": requestID,
	}
	if hasSessionID {
		base["sessionId"] = sessionID
	}

	// Recover from panics (ValidationError, ServerError)
	defer func() {
		if r := recover(); r != nil {
			switch err := r.(type) {
			case *ValidationError:
				result = map[string]interface{}{
					"status": 400,
					"body": mergeMap(copyMap(base), map[string]interface{}{
						"state": "error",
						"error": map[string]interface{}{
							"code":    "VALIDATION_ERROR",
							"message": err.Msg,
						},
					}),
				}
			case *ServerError:
				result = map[string]interface{}{
					"status": err.StatusCode,
					"body": mergeMap(copyMap(base), map[string]interface{}{
						"state": "error",
						"error": map[string]interface{}{
							"code":    err.Code,
							"message": err.Msg,
						},
					}),
				}
			default:
				msg := "Unknown error"
				if e, ok := r.(error); ok {
					msg = e.Error()
				} else if s, ok := r.(string); ok {
					msg = s
				}
				result = map[string]interface{}{
					"status": 500,
					"body": mergeMap(copyMap(base), map[string]interface{}{
						"state": "error",
						"error": map[string]interface{}{
							"code":    "INTERNAL_ERROR",
							"message": msg,
						},
					}),
				}
			}
		}
	}()

	// Validate op is present and a string
	opVal, opExists := envelope["op"]
	if !opExists || opVal == nil {
		return map[string]interface{}{
			"status": 400,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state": "error",
				"error": map[string]interface{}{
					"code":    "INVALID_REQUEST",
					"message": "Missing or invalid 'op' field",
				},
			}),
		}
	}
	op, ok := opVal.(string)
	if !ok || op == "" {
		return map[string]interface{}{
			"status": 400,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state": "error",
				"error": map[string]interface{}{
					"code":    "INVALID_REQUEST",
					"message": "Missing or invalid 'op' field",
				},
			}),
		}
	}

	// Look up operation
	operation, exists := operations[op]
	if !exists {
		return map[string]interface{}{
			"status": 400,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state": "error",
				"error": map[string]interface{}{
					"code":    "UNKNOWN_OP",
					"message": fmt.Sprintf("Unknown operation: %s", op),
				},
			}),
		}
	}

	// Deprecated check -- past sunset date means 410
	if operation.deprecated && operation.sunset != "" {
		sunsetDate, err := time.Parse("2006-01-02", operation.sunset)
		if err == nil && time.Now().After(sunsetDate) {
			var replacement interface{} = nil
			if operation.replacement != "" {
				replacement = operation.replacement
			}
			return map[string]interface{}{
				"status": 410,
				"body": mergeMap(copyMap(base), map[string]interface{}{
					"state": "error",
					"error": map[string]interface{}{
						"code":    "OP_REMOVED",
						"message": fmt.Sprintf("Operation %s has been removed", op),
						"cause": map[string]interface{}{
							"removedOp":   op,
							"replacement": replacement,
						},
					},
				}),
			}
		}
	}

	// Auth check
	if len(operation.authScopes) > 0 {
		authRes := validateAuth(authHeader, operation.authScopes)
		if !authRes.Valid {
			return map[string]interface{}{
				"status": authRes.Status,
				"body": mergeMap(copyMap(base), map[string]interface{}{
					"state": "error",
					"error": map[string]interface{}{
						"code":    authRes.Code,
						"message": authRes.Message,
					},
				}),
			}
		}
	}

	// Idempotency check
	var idempotencyKey string
	if ctx, ok := envelope["ctx"].(map[string]interface{}); ok {
		if ik, ok := ctx["idempotencyKey"].(string); ok {
			idempotencyKey = ik
		}
	}

	if operation.sideEffecting && idempotencyKey != "" {
		idempotencyStoreMu.RLock()
		cached, exists := idempotencyStore[idempotencyKey]
		idempotencyStoreMu.RUnlock()
		if exists {
			return cached.(map[string]interface{})
		}
	}

	// Extract args
	args, _ := envelope["args"].(map[string]interface{})

	// Execute handler
	// Stream operations
	if operation.executionModel == "stream" && operation.streamHandler != nil {
		streamResult := operation.streamHandler(args)
		if streamResult["ok"] == false {
			return map[string]interface{}{
				"status": 200,
				"body": mergeMap(copyMap(base), map[string]interface{}{
					"state": "error",
					"error": streamResult["error"],
				}),
			}
		}
		return map[string]interface{}{
			"status": 202,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state": "streaming",
				"stream": map[string]interface{}{
					"transport": "wss",
					"location":  fmt.Sprintf("/streams/%s", streamResult["sessionId"]),
					"sessionId": streamResult["sessionId"],
					"encoding":  "json",
				},
			}),
		}
	}

	// Async operations
	if operation.executionModel == "async" && operation.asyncHandler != nil {
		asyncResult := operation.asyncHandler(args, requestID)
		if asyncResult["ok"] == false {
			return map[string]interface{}{
				"status": 200,
				"body": mergeMap(copyMap(base), map[string]interface{}{
					"state": "error",
					"error": asyncResult["error"],
				}),
			}
		}
		return map[string]interface{}{
			"status": 202,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state":        "accepted",
				"retryAfterMs": 100,
			}),
		}
	}

	// Sync operations (including media)
	var handlerResult map[string]interface{}
	if operation.acceptsMedia && operation.mediaHandler != nil {
		handlerResult = operation.mediaHandler(args, mf)
	} else if operation.handler != nil {
		handlerResult = operation.handler(args)
	} else {
		return map[string]interface{}{
			"status": 500,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state": "error",
				"error": map[string]interface{}{
					"code":    "INTERNAL_ERROR",
					"message": "No handler for operation",
				},
			}),
		}
	}

	var response map[string]interface{}
	if handlerResult["ok"] == true {
		response = map[string]interface{}{
			"status": 200,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state":  "complete",
				"result": handlerResult["result"],
			}),
		}
	} else {
		// Domain error -- HTTP 200
		response = map[string]interface{}{
			"status": 200,
			"body": mergeMap(copyMap(base), map[string]interface{}{
				"state": "error",
				"error": handlerResult["error"],
			}),
		}
	}

	// Store for idempotency
	if operation.sideEffecting && idempotencyKey != "" {
		idempotencyStoreMu.Lock()
		idempotencyStore[idempotencyKey] = response
		idempotencyStoreMu.Unlock()
	}

	return response
}

func mergeMap(base, overlay map[string]interface{}) map[string]interface{} {
	for k, v := range overlay {
		base[k] = v
	}
	return base
}
